### PR TITLE
make mkDerivation lazier

### PR DESCRIPTION
**Pull request by Profpatsch**
Up until now, it was not possible to access metadata of a package that
was marked as broken (or didn’t pass the license check).
So use cases like creating a list mapping derivation names to derivation
types (which e.g. haskellPackages would need) were not possible and
would simply throw an error, even though there should be no reason to
do so.

This commit draws the brokenness assertion from the toplevel of
mkDerivation into a strict field of the actual derivation call itself.
It also renames `licenseAllowed` to `checkValidity` and makes it a
true function.

That function should probably exposed in `lib/`, along with a function
to recursively filter out reverse dependencies for a broken package.
### Test file

I created [a gist](https://gist.github.com/Profpatsch/a9ade8e229953d31cdfd) to test a trivial usage of mkDerivation. It needs to be put into the `nixpkgs` root folder.
### Implications for nixpkgs
- `nix-env -qa` will not filter out invalid packages anymore (thanks to @vcunat)
- according to @peti, hydra finds out whether a package is marked as broken my accessing the derivation type and checking if that throws an assertion. It needs to be changed to use the actual value of the `meta.broken` attribute.
- ~~nixpkgs will do a complete rebuild~~

cc @vcunat @edolstra @peti

### CODE REVIEW

**Comment by peti** [comment]
For reference: the motivation for this PR is https://github.com/NixOS/nixpkgs/issues/7541 and https://github.com/NixOS/cabal2nix/issues/130. https://github.com/NixOS/nixpkgs/pull/9305 Was a previous attempt to fix this issue.

**Comment by Mathnerd314** [comment]
As mentioned in #9305, Hydra can be addressed by settings [meta](https://github.com/Profpatsch/nixpkgs/commit/64f598fd559e354e02e3d9aba5343d4ce5b79ae5#diff-3c8a88619c3ca3d61eb0ce8794d9bb74R204).hydraPlatforms = [] if the derivation isn't valid.

**Comment by peti** [comment]
The patch causes no mass-rebuild, fortunately; no store paths seem affected.

**Comment by vcunat** [comment]
Ah, right, it seems so from the changes. I got mislead by the claim in the description (so I crossed it out now).

**Comment by vcunat** [comment]
@Mathnerd314: this isn't just about Hydra. We also have `nix-env -qa` that lists "available packages", and it seems precisely the same problem &ndash; we want to get every package except those that are filtered by some condition that we set, e.g. depending on something with unfree license.

**Comment by Profpatsch** [comment]
Ah, it doesn’t rebuild because that part is in the topmost layer, not the innermost, right?

@vcunat As you said, we need to know the reverse dependencies to filter invalid packages. For reference, [here](https://hackage.haskell.org/package/packdeps-0.4.1/docs/src/Distribution-PackDeps.html#getReverses) is the algorithm how they do it for Hackage on http://packdeps.haskellers.com/

**Comment by Profpatsch** [comment]
Ah @Mathnerd314, you already proposed a similar patch. But it has more implications for the whole infrastructure, as mentioned above.

**Comment by domenkozar** [comment]
+1 cc @nbp

**Comment by domenkozar** [comment]
Are there any drawbacks to this approach? I really see none.

**Comment by Profpatsch** [comment]
@domenkozar None afaik. But we should implement a function for reverse dependency tracking in order to keep the current functionality (of not building packages depending on invalid packages).

File: pkgs/stdenv/generic/default.nix [hunk 1]
```diff
@@ -111,30 +111,36 @@ let
...
+        throw ("Package ‘${attrs.name or "«name-missing»"}’ in ${pos''} ${errormsg}, refusing to evaluate."
+        + (lib.strings.optionalString (reason != "blacklisted") ''
 
           For `nixos-rebuild` you can set
-            { nixpkgs.config.allow${unfreeOrBroken} = true; }
+            { nixpkgs.config.allow${reason} = true; }
           in configuration.nix to override this.
           For `nix-env` you can add
-            { allow${unfreeOrBroken} = true; }
+            { allow${reason} = true; }
```
Comment (nbp) [inline pkgs/stdenv/generic/default.nix]: These messages use to have capitalized letters for `Unfree` and `Broken`, this is no longer the case, and this contradict the documentation and the rest of the check for these flags.

If this is a deliberate change, then this would need additional documentation and code changes, otherwise this should be fixed before landing.

**Comment by nbp** [comment]
Except for the previous remark, this sounds like a good approach.

File: pkgs/stdenv/generic/default.nix (hunk 1 above)
Comment (Profpatsch) [inline pkgs/stdenv/generic/default.nix]: Yes, this was intentional, as some were written capitalized and some where not. I thought they are just used internally and it was a simple inconsistency. Which docs do I need to change?

Comment (Mathnerd314) [inline pkgs/stdenv/generic/default.nix]: This is in the error message shown when you try to install an unfree package, like https://github.com/NixOS/nixpkgs/issues/4766. With this patch it says `` { nixpkgs.config.allowbroken = true; } ``, instead of the typical ``allowBroken``.

**Comment by peti** [comment]
This PR mixes two things into one patch: there is (1) a general re-factoring that's uncontroversial, IMHO, and that could me merged today. Then, there is (2) a change to the semantics of `mkDerivation` that is highly desirable, but difficult to get right and that needs changes to other tools before it can be merged. Personally, I'd like to see the (1) part factored out into a separate PR that we can merge now and (2) re-based on top of that patch. That would give us the benefit of (1) regardless of whether (2) is going to be merged or not, and it should dramatically simplify the patch complexity of (2), which makes it easier to review and easier for people to understand.

Generally speaking, we cannot merge (2) until we have the necessary companion patches for Hydra and `nix-env` available. The change to `nix-env` in particular forces a bit of a delay on us because we'll need the new version of Nixpkgs that contains the sematic change from (2) to depend on Nix 1.11 as old versions can no longer handle the package database correctly after the semantics of `mkDerivation` have changed. The required change to Hydra is less critical as Hydra is an unreleased "use at your own risk" kind-of program, and people who run it are expected to be able to cope with these kind of changes. Still, Hydra needs to be adapted to deal correctly with packages that set `meta.broken = true`.

**Comment by vcunat** [comment]
Personally, I've got no idea how to sensibly reinstate the license/brokenness/etc. checks in all dependencies without throwing on evaluation (which this PR removes).

File: pkgs/stdenv/generic/default.nix (hunk 1 above)
Comment (Profpatsch) [inline pkgs/stdenv/generic/default.nix]: Well, then we need to change that code, it’s kind of silly to construct documentation of symbols from a dynamic and fixed string.

**Comment by Profpatsch** [comment]
@Mathnerd314 I introduced a small local function that uppercases the first character. Should work now, if I didn’t miss something else.

**Comment by vcunat** [comment]
This seems interesting: https://github.com/NixOS/nix/pull/721#issuecomment-160263390.

**Comment by Profpatsch** [comment]
@vcunat So we should introduce yet another magic attribute and not fix this? I’m not very sold on that idea. 

Relying on the short-circuiting, partly magic abilities of `throw` (C++ code can skip errors) instead of doing it the right way (calculating reverse dependencies) at the very core of nixpkgs seems like a bad call.

**Comment by shlevy** [comment]
@Profpatsch The attribute has uses besides this one, and unlike the solution here will actually preserve the brokenness propagation.

**Comment by Profpatsch** [comment]
@shlevy By (ab-)using the strict properties of throw? Will I be able to access e.g. <drv>.name?

**Comment by shlevy** [comment]
No, by delaying the call to mkDerivation, which itself will throw. foo.name will evaluate just fine, but ${foo} will not.

**Comment by shlevy** [comment]
Note that we already have magic attributes anyway: `${ { outPath = "foo"; } }` works today and is really just a special case of `__toString`

**Comment by Ericson2314** [comment]
Oh yuck, not even a `__` or something to hint that it is magic.

**Comment by shlevy** [comment]
Sure, but it's better than writing `${drv.outPath}` everywhere we put `${drv}` now :grinning:

**Comment by zimbatm** [comment]
Has this work been abandoned ? Anything I can do to help ?

**Comment by shlevy** [comment]
FWIW I still maintain that an approach that uses `__toString` would be better here.

**Comment by Profpatsch** [comment]
@zimbatm `nix-env` is rather sloppy and uses the asserts to stop search for reverse dependencies. So in order to make the patch work we need to implement a function that calculates reverse dependencies. The [Haskell reverse dependency monitor](https://hackage.haskell.org/package/packdeps-0.4.2.1/docs/Distribution-PackDeps.html) does exactly this.

**Comment by peti** [comment]
I suppose it's fair to say that this PR is well and truly stuck. I don't see any practical way forward. :disappointed:

**Comment by Profpatsch** [comment]
@peti I’m sorry, I should have just implemented the function. I still intent to do this, but it hasn’t been on my priority list.

**Comment by abbradar** [comment]
FWIW, yet another use case: https://github.com/NixOS/nixpkgs/issues/14890
tl;dr: use `passthru` attributes from a non-valid (for some reason) derivation.

**Comment by domenkozar** [comment]
Just curious, if I access `drv.name` with this patch, will derivation be used lazily? If so, very cool :)

**Comment by knedlsepp** [comment]
This behavior also limits in the following example:
The `openmpi` package has a (in theory optional) dependency on the infiniband library `libibverbs`, which is only supported on linux and freebsd. I would like to build `openmpi` on darwin, so I was trying to change the `openmpi` package to only depend on `libibverbs` if my platform supports it:
[`++ optional (any (x: x == stdenv.system) libibverbs.meta.platforms) libibverbs;`](https://github.com/knedlsepp/nixpkgs/commit/14ad1af207a7acadb7ed3eb027de9d731dc8860d#diff-551877b1e2e1f35a39c35eb0241c085aR30)
This however does not work and thus I will probably have to hard-code it via:
`++ optional (stdenv.isLinux || stdenv.isFreeBSD) libibverbs;`

**Comment by peti** [comment]
This PR appears to be dead in the water. :disappointed:

**Comment by Profpatsch** [comment]
@peti I wouldn’t lose hope yet, there’s multiple hackathons coming up and I’ve already thought about how the reverse dependency function needs to look.

**Comment by vcunat** [comment]
Another shot to improve this: #22277.
